ibc/store/provable_store.go

```diff
@@ -0,0 +1,244 @@
+	root, nodeStore := p.bus.GetTreeStore().GetTree(trees.IBCStateTree)
+	lazy := smt.ImportSparseMerkleTree(nodeStore, root, sha256.New())
+	return createNonMembershipProof(lazy, prefixed)
+	**/
+	return nil, nil
+}
+
+// Set updates the persistence layer with the new key-value pair at the latest height and
+// emits an UpdateIBCStore event to the bus for propagation throughout the network, to be
+// included in each node's mempool and thus the next block
+func (p *provableStore) Set(key, value []byte) error {
+	prefixed := applyPrefix(p.prefix, key)
+	currHeight := int64(p.bus.GetConsensusModule().CurrentHeight())
+	rwCtx, err := p.bus.GetPersistenceModule().NewRWContext(currHeight)
+	if err != nil {
+		return err
+	}
+	defer rwCtx.Release()
+	if err := rwCtx.SetIBCStoreEntry(prefixed, value); err != nil {
+		return err
+	}
+	p.cache = append(p.cache, &cachedEntry{
+		storeName:   p.name,
+		height:      uint64(currHeight),
+		prefixedKey: prefixed,
+		value:       value,
+	})
+	// TODO(#854): Implement emit functions
+	// return emitUpdateStoreEvent(p.prefix, key, value)
+	return nil
+}
+
+// Delete updates the persistence layer with the key and nil value pair at the latest height
+// and emits an PruneIBCStore event to the bus for propagation throughout the network, to be
+// included in each node's mempool and thus the next block
+func (p *provableStore) Delete(key []byte) error {
+	prefixed := applyPrefix(p.prefix, key)
+	currHeight := int64(p.bus.GetConsensusModule().CurrentHeight())
+	rwCtx, err := p.bus.GetPersistenceModule().NewRWContext(currHeight)
+	if err != nil {
+		return err
+	}
+	defer rwCtx.Release()
+	if err := rwCtx.SetIBCStoreEntry(prefixed, nil); err != nil {
+		return err
+	}
+	p.cache = append(p.cache, &cachedEntry{
+		storeName:   p.name,
+		height:      uint64(currHeight),
+		prefixedKey: prefixed,
+		value:       nil,
+	})
+	// TODO(#854): Implement emit functions
+	// return emitPruneStoreEvent(p.prefix, key)
+	return nil
+}
+
+// GetCommitmentPrefix returns the CommitmentPrefix for the store
+func (p *provableStore) GetCommitmentPrefix() coreTypes.CommitmentPrefix { return p.prefix }
+
+// Root returns the current root of the IBC state tree
+func (p *provableStore) Root() ics23.CommitmentRoot {
+	// TODO(#854): Implement tree retrieval
+	/**
+	root, _ := p.bus.GetTreeStore().GetTree(trees.IBCStateTree)
+	return root
+	**/
+	return nil
+}
+
+// CacheEntries writes all local changes to disk and clears the in-memory cache
+func (p *provableStore) CacheEntries(store kvstore.KVStore) error {
+	for _, entry := range p.cache {
+		key, value := entry.prepare()
+		if err := store.Set(key, value); err != nil {
+			return err
+		}
+	}
+	p.cache = make([]*cachedEntry, 0)
+	return nil
+}
+
+// PruneCache removes all entries from the cache at the given height
+func (p *provableStore) PruneCache(store kvstore.KVStore, height uint64) error {
+	keys, _, err := store.GetAll([]byte(fmt.Sprintf("%s/%d", p.name, height)), false)
```

Comment:
Have you thought about about:

1. Updating `PersistenceModule` to allowing returning the writeContext (it's a singleton initialized by NewRWContext)

2. Using the write context for the cache? That way as it errors or is released, the logic is centralized in one place.

cc @dylanlott for thoughts on this as well

Reply:
A little unsure of what you mean. Having a new cache table for example in the postgres DB and using that as our cache? The idea behind the cache is that if the store change event fails to propagate or the node crashes we can retry. I assumed an in memory cache which is periodically written to disk and pruned (handled by store manager soon TM) would be better but happy to discuss this further.

Reply:
added at top of this file